core/modules/replica/AbortTransactionJob.cc

```diff
@@ -0,0 +1,344 @@
+                _worker2tables[worker].push_back(table + "_" + to_string(chunk));
+            }
+        }
+    }
+
+    // Submit the first batch of requests. The number of the requests not to exceed
+    // the processing capacity of all workers.
+
+    size_t const maxRequests =
+        _workers.size() * controller()->serviceProvider()->config()->workerNumProcessingThreads();
+
+    if (0 == _submitNextBatch(lock, maxRequests)) {
+        // No tables to be processed
+        finish(lock, ExtendedState::SUCCESS);
+    }
+}
+
+
+void AbortTransactionJob::cancelImpl(util::Lock const& lock) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << __func__);
+
+    for (auto&& request: _requests) {
+        request->cancel();
+    }
+}
+
+
+void AbortTransactionJob::notify(util::Lock const& lock) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << __func__);
+
+    notifyDefaultImpl<AbortTransactionJob>(lock, _onFinish);
+}
+
+
+void AbortTransactionJob::_onRequestFinish(SqlDeleteTablePartitionRequest::Ptr const& request) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << __func__
+         << "  worker=" << request->worker()
+         << " _requests.size()=" << _requests.size()
+         << " _numLaunched=" << _numLaunched
+         << " _numFinished=" << _numFinished
+         << " _numSuccess=" << _numSuccess);
+
+    if (state() == State::FINISHED) return;
+
+    util::Lock lock(_mtx, context() + string(__func__));
+
+    if (state() == State::FINISHED) return;
+
+    ++_numFinished;
+    if (request->extendedState() == Request::SUCCESS) ++_numSuccess;
+
+    // Replace the finished job with a new one unless the input set is empty
+    if (0 == _submitNextBatch(lock, 1)) {
+
+        if (_numFinished == _numLaunched) {
+
+            LOGS(_log, LOG_LVL_DEBUG, context() << __func__
+                 << "  worker=" << request->worker()
+                 << " _requests.size()=" << _requests.size());
+
+            // Harvest results from all jobs (regardless of their completion status)
+            // and be done.
+
+            for (auto&& ptr: _requests) {
+                LOGS(_log, LOG_LVL_DEBUG, context() << __func__
+                     << " requestId=" << ptr->id()
+                     << " targetRequestParams=" << ptr->targetRequestParams());
+                auto const table = ptr->targetRequestParams().table;
+                _resultData.completed [ptr->worker()][table] = ptr->extendedState() == Request::SUCCESS;
+                _resultData.resultSets[ptr->worker()][table] = ptr->responseData();
+            }
+            finish(lock, _numSuccess == _numLaunched ? ExtendedState::SUCCESS
+                                                     : ExtendedState::FAILED);
+        }
+    }
+}
+
+
+size_t AbortTransactionJob::_submitNextBatch(util::Lock const& lock,
+                                             size_t const maxRequests) {
+
+    if (0 == maxRequests) return maxRequests;
```

Comment:
This may happen only if there are no workers configured as participating in the Replication/Ingest operations at a given moment of time. For example, if the workers were brought down. And it would be undesired to shut down the whole Master Replication Controller service because of that. The service keeps re-evaluating worker statuses, and it may also receive a command for adding a new worker (or reactivating the one which was brought down). So, that the operation could be attempted.  Hence, the current algorithm tried to keep the system stable.

Reply:
So no enabled nodes, nothing to do. It might be more clear to check for maxRequests==0 before calling _submiutNextBatch in startImpl. Or a comment here.

Reply:
Agreed

Reply:
Having thought more on this, I'm taking my word back. I'd rather keep this logic as it's now because there may be 3 reasons why method `_submitNextBatch` may return `0`:
# if the number of workers in the setup is `0`
# if a value o the Configuration parameter `workerNumProcessingThreads` is set to `0`
# if no chunks to be cleaned up uploaded by a client
I think it's better to have just one call to `_submitNextBatch` and one test for its result to be `0` rather than expanding the body of the following method to test each of these conditions and take the same action, like shown below
```
void AbortTransactionJob::startImpl(util::Lock const& lock) {
    ..
    if (0 == workers.size()) {
        finish(lock, ExtendedState::SUCCESS);
    }
    if (0 == controller()->serviceProvider()->config()->workerNumProcessingThreads()) {
        finish(lock, ExtendedState::SUCCESS);
    }
    size_t const maxRequests =
        _workers.size() * controller()->serviceProvider()->config()->workerNumProcessingThreads();

    if (0 == _submitNextBatch(lock,maxRequests)) {
        finish(lock, ExtendedState::SUCCESS);
    }

```